### PR TITLE
Deprecate use of private variables from other modules

### PR DESCRIPTION
**Pull request by RazvanN7**
Since #6660 (the work of the devil) was reverted, this PR implements the same feature except this time a deprecation message is issued and not an error

### CODE REVIEW

File: src/ddmd/dimport.d
Comment (mathias-lang-sociomantic) [inline src/ddmd/dimport.d]: I'd rephrase as something like "Accessing private member FQN is deprecated"

**Review by mathias-lang-sociomantic — DISMISSED** [review]
Needs test and release notes entry

File: src/ddmd/dimport.d
Comment (RazvanN7) [inline src/ddmd/dimport.d]: How does "module $module member $name is not accessible because it is private". I think having the word "deprecated" in the message is redundant since the phrase starts with "Deprecation: ".

**Comment by RazvanN7** [comment]
@mathias-lang-sociomantic I've never done a release entry. Could you guide me with the process of doing it?

Thanks

File: src/ddmd/dimport.d
Comment (mathias-lang-sociomantic) [inline src/ddmd/dimport.d]: Sounds good :+1:

**Comment by mathias-lang-sociomantic** [comment]
@RazvanN7 : https://github.com/dlang/dmd/tree/master/changelog

**Comment by RazvanN7** [comment]
Any ideas on how to solve the continous-integration/jenkins error in vibe.d ? I see that the deprecation message causes some linking errors and I think that vibe.d should be updated to not use/import private members from other modules

**Comment by ZombineDev** [comment]
> Any ideas on how to solve the continous-integration/jenkins error in vibe.d ?

No need to worry, these linker errors are not caused by your pull request, see https://github.com/dlang/dmd/pull/6682#issuecomment-292773861

**Comment by ibuclaw** [comment]
Shouldn't the mechanism for bug34 catch this? Also what about protected and package?

**Comment by RazvanN7** [comment]
@ibuclaw I don't think that protected and package have anything to do with this PR. anyway, how we move forward from here?

**Comment by ibuclaw** [comment]
If they suffer from the same bug, then they must be fixed also.

**Review by andralex — APPROVED** [review]

**Review by ibuclaw — DISMISSED** [review]
1. I think the fix should be in access.d, not where you currently have it.

2. Consider the following example for fail15896:
```
import imp15896 : publicbar, privatebar, packagebar;
int func()
{
    publicbar += 1;
    privatebar += 1;
    packagebar +=1;
    return 0;
}
```
This should throw a compile-time error where marked with NG below:
```
int bar = 4;
public ref int publicbar()       { return bar; }  // OK
private ref int privatebar()     { return bar; }  // NG
package ref int packagebar()     { return bar; }  // NG
```
Therefore this PR should also ensure the same behaviour for variables:
```
public int publicbar = 4;   // OK
private int privatebar = 4; // OK (but should be NG)
package int packagebar = 4; // OK (but should be NG)
```

**Comment by RazvanN7** [comment]
@ibuclaw That is what this PR does for private members. I think that package should be the subject of a different PR.

**Comment by ibuclaw** [comment]
It's the same bug, so they should be fixed together.

**Comment by andralex** [comment]
@MartinNowak should decide on this. Do we fix all in one deprecation shot, or do we stagger fixes?

**Comment by andralex** [comment]
ping @MartinNowak @WalterBright

**Comment by andralex** [comment]
@RazvanN7 please rebase

File: src/ddmd/dimport.d
Comment (ibuclaw) [inline src/ddmd/dimport.d]: I think it would be better if you used one of these functions.

https://github.com/dlang/dmd/blob/8d6efa5dd461ac9c28caf1964a9dab2ec86232b3/src/dmd/access.d#L481-L525

Comment (ibuclaw) [reply src/ddmd/dimport.d]: Either:
```
if (!symbolIsVisible(mod, importedSymbol) { error; }
```
or
```
if (!symbolIsVisible(sc, importedSymbol) { error; }
```

**Comment by RazvanN7** [comment]
@ibuclaw @mathias-lang-sociomantic I rebased and updated the code. Is this cool now?

File: changelog/private.dd
Comment (ibuclaw) [inline changelog/private.dd]: Will be transformed?

Comment (ibuclaw) [reply changelog/private.dd]: Then again the wording in general could be improved.

**Review by ibuclaw — COMMENTED** [review]
Change seems fine.

File: changelog/private.dd
Comment (JinShil) [reply changelog/private.dd]: Proposed wordsmithing:

> $(LINK2 $(ROOT_DIR)spec/attribute.html#visibility_attributes, The specification states) that a private member is visible on from within the same module.  Prior to this release, due to a bug, some imported private members were visible from other modules, violating the specification.  Beginning with this release, accessing said private members from another module will result in a deprecation message which has been scheduled to be changed to an error in approximately 1 year.

File: src/dmd/dsymbolsem.d
Comment (JinShil) [inline src/dmd/dsymbolsem.d]: Please add the following comments to make it easier to follow up on changing it to an error.

```D
// @@@DEPRECATED_2018-01@@@
// Deprecation for 1 year after which this should be changed to an error
```

**Review by JinShil — CHANGES_REQUESTED** [review]

**Comment by JinShil** [comment]
Is there a bug report for this?

File: src/dmd/dsymbolsem.d
Comment (wilzbach) [reply src/dmd/dsymbolsem.d]: Shouldn't this be 2019?

Comment (JinShil) [reply src/dmd/dsymbolsem.d]: It was deprecated in 2018-01. It will be changed to an error in 2019-01.  I read the date as the date is was deprecated, not the date it should be changed to an error.

Comment (wilzbach) [reply src/dmd/dsymbolsem.d]: That's not how we use this pattern at Phobos. The string is intended to be greppable in the future, i.e. you can do `grep "@@@DEPRECATED_2019"` for everything that needs maintenance action in 2019 and `grep "@@@DEPRECATED_2019_01"` for everything within a specific month.

File: changelog/private.dd
Comment (RazvanN7) [reply changelog/private.dd]: Thanks. I copy-pasted your words.

**Comment by RazvanN7** [comment]
@JinShil There was a bug report which closed when I fixed the issue to issue an error. Later when the PR was reverted so that the error could be transformed into a deprecation the bug report was no longer opened. It's this one : https://issues.dlang.org/show_bug.cgi?id=15896. If this PR is merged then we will be ok.

File: src/dmd/dsymbolsem.d
Comment (JinShil) [reply src/dmd/dsymbolsem.d]: That doesn't seem right to me, but I'll play along.  It probably needs to be disambiguated somehow.  Please offer @RazvanN7 a suggestion so we can get this merged.

**Comment by RazvanN7** [comment]
ping @mathias-lang-sociomantic @ibuclaw @JinShil

File: src/dmd/dsymbolsem.d
Comment (wilzbach) [reply src/dmd/dsymbolsem.d]: Well it's the existing practice to document deprecations, see e.g. https://github.com/wilzbach/phobos/blob/complex/std/datetime/timezone.d#L145

How else are you going to find them? I already did -> Just change to 2018 -> 2019

File: src/dmd/dsymbolsem.d
Comment (wilzbach) [inline src/dmd/dsymbolsem.d]: As I tried to explain to @JinShil, this needs to be `2019` instead of `2018` as it should be the date when the maintenance act is needed.
See e.g. https://github.com/wilzbach/phobos/blob/complex/std/datetime/timezone.d#L145

Comment (JinShil) [reply src/dmd/dsymbolsem.d]: Also, please add some explanatory text such.  Something like `Deprecated in 2018-01.  Change to error in 2019-01.  @@@DEPRECATED_2019-01@@@`.

**Comment by JinShil** [comment]
> There was a bug report which closed when I fixed the issue to issue an error. Later when the PR was reverted so that the error could be transformed into a deprecation the bug report was no longer opened.

OK, do we need to change a commit message to "Fix 15896..." so the automation will add a link to the changelog?

**Comment by wilzbach** [comment]
> OK, do we need to change a commit message to "Fix 15896..." so the automation will add a link to the changelog?

https://issues.dlang.org/show_bug.cgi?id=15896 is already marked as RESOLVED FIXED. It could be REOPENED and tagged here, but we already have a [changelog entry](http://dtest.dlang.io/artifact/website-0c475b6499fd5fdbcfb1961b6513f0a8af232c56-8a4e23642e5434228e086a83c4849bc3/web/changelog/pending.html#private) which is far more valuable than a Bugzilla reference.
-> It wouldn't hurt, but not a requirement either (in this special case)

File: src/dmd/dsymbolsem.d
Comment (RazvanN7) [reply src/dmd/dsymbolsem.d]: Done, thanks.

File: src/dmd/dsymbolsem.d
Comment (JinShil) [inline src/dmd/dsymbolsem.d]: Sorry, but I found one more thing.  We need to put tick marks around the %s.  (e.g. \`%s\`).  Remove the single quotes.

**Review by JinShil — CHANGES_REQUESTED** [review]

File: src/dmd/dsymbolsem.d
Comment (JinShil) [reply src/dmd/dsymbolsem.d]: And don't forget to update the error messages in the test case too.

**Review by JinShil — APPROVED** [review]

File: src/dmd/dsymbolsem.d
Comment (RazvanN7) [reply src/dmd/dsymbolsem.d]: Oh, missed that one too. Thanks!

**Comment by JinShil** [comment]
Looks like a legitimate failure for dstats.  We may either need to modify the test, or modify the source code to get this merged.

**Comment by wilzbach** [comment]
> Looks like a legitimate failure for dstats and experimental/color in Jenkins (at least if I'm reading the error report properly. Are we treating deprecations as errors in our tests?).

@RazvanN7 you are in luck. This is needed for https://github.com/dlang/dmd/pull/7241 too. The PR is already out:

https://github.com/DlangScience/dstats/pull/39

**Comment by andralex** [comment]
@wilzbach who is on hook with dstats? Can you please send an email?

**Comment by andralex** [comment]
@ibuclaw please take one more look and possibly approve?

**Comment by wilzbach** [comment]
> @wilzbach who is on hook with dstats?

I think that would be @John-Colvin 

> Can you please send an email?

OK. I just did so.
We two options for unmaintained projects:
- move it to dlang-community (this depends on someone with admin permissions to dstats being active once as otherwise the DUB package won't direct to the right repo)
- remove it from the Project Tester (please mind that ggplotd depends on dstats, so we would have to remove that one too)

I would prefer option 1.
Pinging other members from DlangScience who might have merge/admin rights: @9il @BlackEdder @dataPulverizer @kyllingstad

**Comment by 9il** [comment]
I have rights but John and Lars should decide.

**Comment by dataPulverizer** [comment]
I have rights on DlangScience, but I am the wrong person to be approving things. I'm happy to give rights to more suitable person.

**Comment by wilzbach** [comment]
@RazvanN7 the ggplotd timeout is a true positive.
It's easy to reproduce:

```bash
git clone https://github.com/BlackEdder/ggplotd
cd ggplotd
dub test --compiler=/home/seb/dlang/dmd/generated/linux/release/64/dmd
```

File: test/fail_compilation/fail15896.d
Comment (ibuclaw) [inline test/fail_compilation/fail15896.d]: Looks like you need to highlight the module name now.

**Review by ibuclaw — DISMISSED** [review]

File: test/fail_compilation/fail15896.d
Comment (RazvanN7) [reply test/fail_compilation/fail15896.d]: done.

**Review by MartinNowak — CHANGES_REQUESTED** [review]
Please don't merge before I reviewed this.

**Comment by MartinNowak** [comment]
I don't quite follow the action in https://github.com/BlackEdder/ggplotd/pull/45, seems like noone figured out why the compiler runs into an infinite loop.

Clearly this is not allowed
```d
alias T = int;
alias T = T;
```
and so
```d
import std.stdio : writeln;
alias writeln = writeln;

void main()
{
    writeln("Hello");
}
```
should print an error about writeln aliasing itself, but it currently compiles.
In particular because it is syntactially very close to how you construct overload sets, this is likely sth. that people will try from time to time.
https://dlang.org/hijack.html#overload-sets

Unfortunately there is no explanation for the infinite loop introduced with **this** PR.
Just fixing code that triggers the bug to pass CI, seems like the wrong action here.

I can reproduce an ICE for the following similar case.
```d
import std.stdio : writeln;
import bar;
alias writeln = writeln; // <- should error, as the selectively imported writeln already is an alias
alias writeln = bar.writeln; // OK, overload set of selectively imported writeln and bar.writeln

void main()
{
    writeln();
}
```

This is the profile I get from the inifite loop in the compiler.
```
59.12%  dmd         dmd                  [.] _D3dmd6access19mostVisibleOverloadFCQBi7dsymbol7DsymbolZQv
10.33%  dmd         dmd                  [.] AliasDeclaration::isOverloadable
 8.67%  dmd         dmd                  [.] AliasDeclaration::toAlias
 5.72%  dmd         dmd                  [.] Prot::isMoreRestrictiveThan
 3.92%  dmd         dmd                  [.] Declaration::prot
 3.40%  dmd         dmd                  [.] Dsymbol::isFuncAliasDeclaration
 1.74%  dmd         dmd                  [.] Dsymbol::isFuncDeclaration
 1.48%  dmd         dmd                  [.] OverDeclaration::isOverDeclaration
 1.48%  dmd         dmd                  [.] Dsymbol::isTemplateDeclaration
 1.03%  dmd         dmd                  [.] AliasDeclaration::isAliasDeclaration
 0.72%  dmd         dmd                  [.] Dsymbol::isOverDeclaration
 0.71%  dmd         dmd                  [.] OverDeclaration::isOverloadable
 0.70%  dmd         dmd                  [.] Dsymbol::toAlias
```

File: changelog/private.dd
Comment (MartinNowak) [inline changelog/private.dd]: `s/some imported private members/private selective imports/`
Also just mention that it has been deprecated, we have a general schedule for deprecations (6 months), but deviate from that when it seems too disturbing.

File: test/fail_compilation/fail15896.d
Comment (MartinNowak) [inline test/fail_compilation/fail15896.d]: s/accessible/visible/
Accessibility is the old protection scheme that will soon be removed.
Also check the other visibility messages, they should be more sth. like `X of mod M is not visible from mod N` or so.

File: src/dmd/dsymbolsem.d
Comment (MartinNowak) [inline src/dmd/dsymbolsem.d]: The comment was added with https://github.com/dlang/dmd/pull/7760.

We should at some point change the logic from pulling out a symbol and then checking it's visibility, to passing the visibility level to search and reduce it with according to import protection (like a funnel).
This would also be necessary to fix `package import pkg.common;`, but requires quite some changes and partial order of visibility.

**Review by MartinNowak — APPROVED** [review]
Looks OKish, only just found the other half in #7760, please fix the infinite loop though.

**Comment by RazvanN7** [comment]
@MartinNowak The infinte loop introduced by this PR in ggplotd is due to the function "mostVisibleOverload" which I did not touch. The problem here is, as you have stated, the fact that the compiler allows `alias ApplyLeft = ApplyLeft`. I wanted to move on with this (which is a separate issue) and then investigate the recursive alias problem.

**Comment by andralex** [comment]
@MartinNowak @ibuclaw anything else to do here?

**Comment by JinShil** [comment]
Restarting Jenkins due to https://github.com/dlang/ci/pull/151

**Comment by wilzbach** [comment]
That usually doesn't help. Often you need to give Jenkins a git commit sha to reload the config - though in this case it might work...
(In my experience if the dlang/ci PR has been made at the same day -> needs new git Sha for a real restart)

**Comment by MartinNowak** [comment]
> @MartinNowak The infinte loop introduced by this PR in ggplotd is due to the function mostVisibleOverload in dmd/access.d which I did not touch.

But you did call it in a situation where it wasn't previously called, that counts as introducing a bug ;).

@RazvanN7 I think the way forward here is very clear and simple, make
```d
import std.stdio : writeln;
alias writeln = writeln;
```
an error (maybe even without deprecation as it seems extremely rare in commited code).
Error message should be sth. like `recursive alias, use a qualified name to create overload sets`.

The reason why should fix this in advance is because overload sets are rather complex, and this looks like a plausible syntax to combine the selectively imported `writeln` into an overload set named `writeln`.

I'll try to work on a PR later today, but at latest for the 2.079.0 beta on 2018-02-14, so that we don't release with that bug.

**Comment by MartinNowak** [comment]
> That usually doesn't help. Often you need to give Jenkins a git commit sha to reload the config - though in this case it might work...
> (In my experience if the dlang/ci PR has been made at the same day -> needs new git Sha for a real restart)

To clarify this, by design, hitting replay on Jenkins checks out the same repo versions that were used last time. This does only apply to those repositories for which we use Jenkin's internal checkout mechanism.
That includes the ci repo with the [runPipeline.groovy](https://github.com/dlang/ci/blob/4602173ee34bda6a29392c3db4768af1db346221/vars/runPipeline.groovy) script and the repo that triggered the build.
So hitting replay won't pick up any changes for those 2 repos, filed as https://github.com/dlang/ci/issues/160.

**Comment by wilzbach** [comment]
(closed and reopened for Semaphore to know about this PR)

**Comment by wilzbach** [comment]
@MartinNowak applied your comments to the changelog and fixed a few other nits in the changelog message.

**Comment by timotheecour** [comment]
wrote https://github.com/dlang/dmd/pull/7930 to fix issue with alias X=X making compiler hang

**Pull request merged**
